include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -567,65 +567,66 @@ class TypeRefBuilder {
-    if (!maybeGenericParamsPerLevel)
-      return nullptr;
-
-    auto genericParamsPerLevel = *maybeGenericParamsPerLevel;
-
-    auto kind = node->getKind();
-    // Kinds who have a "BoundGeneric..." variant.
-    if (kind != Node::Kind::Class && kind != Node::Kind::Structure &&
-        kind != Node::Kind::Enum)
-      return nullptr;
-    auto mangling = Demangle::mangleNode(node);
-    if (!mangling.isSuccess())
-      return nullptr;
-
-    if (shapeIndex >= genericParamsPerLevel.size())
-      return nullptr;
-
-    auto numGenericArgs = genericParamsPerLevel[shapeIndex];
-
-    auto startOffsetFromEnd = argsIndex + numGenericArgs;
-    auto endOffsetFromEnd = argsIndex;
-    if (startOffsetFromEnd > args.size() || endOffsetFromEnd > args.size())
-      return nullptr;
-
-    std::vector<const TypeRef *> genericParams(
-        args.end() - startOffsetFromEnd, args.end() - endOffsetFromEnd);
-
-    const BoundGenericTypeRef *parent = nullptr;
-    if (node->hasChildren()) {
-      // Skip over nodes that are not of type class, enum or struct
-      auto parentNode = node->getFirstChild();
-      while (parentNode->getKind() != Node::Kind::Class &&
-             parentNode->getKind() != Node::Kind::Structure &&
-             parentNode->getKind() != Node::Kind::Enum) {
-        if (parentNode->hasChildren()) {
-          parentNode = parentNode->getFirstChild();
-        } else {
-          parentNode = nullptr;
-          break;
-        }
+  // Construct a bound generic type ref along with the parent type info
+  // The parent list contains every parent type with at least 1 generic
+  // type parameter.
+  const BoundGenericTypeRef *reconstructParentsOfBoundGenericType(
+      const NodePointer startNode,
+      const std::vector<size_t> &genericParamsPerLevel,
+      const llvm::ArrayRef<const TypeRef *> &args)
+  {
+    // Collect the first N parents that potentially have generic args
+    // (Ignore the last genericParamPerLevel, which
+    // applies to the startNode itself.)
+    std::vector<NodePointer> nodes;
+    NodePointer node = startNode;
+    while (nodes.size() < genericParamsPerLevel.size() - 1) {
+      if (!node || !node->hasChildren()) {
+	return nullptr;
       }
-      if (parentNode) {
-        if (shapeIndex > 0)
-          parent = createBoundGenericTypeReconstructingParent(
-              parentNode, decl, --shapeIndex, args, argsIndex + numGenericArgs);
-        else
-          return nullptr;
+      node = node->getFirstChild();
+      switch (node->getKind()) {
+      case Node::Kind::Class:
+      case Node::Kind::Structure:
+      case Node::Kind::Enum:
+	nodes.push_back(node);
+	break;
+      default:
+	break;
       }
     }
+    assert(nodes.size() == genericParamsPerLevel.size() - 1);
+
+    // We're now going to build the type tree from the
+    // outermost parent in, which matches the order of
+    // the generic parameter list and genericParamsPerLevel.
+    std::reverse(nodes.begin(), nodes.end());
```

Comment:
Above, we walk the parent list from the innermost to outermost.  That is, for `Foo<Int>.Bar<Double>.Baz<String>.Quux`, we visit `Baz`, then `Bar`, then `Foo` (skipping `Quux`, since that's the start node, not one of the parents we're collecting here.

I then need to build the new list in the opposite order, which is why I reverse the list here.